Made request.body parsing more robust

radiantdinosaurs opened the pull request:
When validating file format, validateFileFormat (in write-file) analyzes the file signature (i.e., 'magic numbers') encoded as base64 rather than checking the 'data label.' This means the user no longer needs to include a data label when sending base64. Also, this means we no longer need a parser to strip the data label from base64 strings, so I removed the base64 parse file and related tests.

radiantdinosaurs committed:
Made request.body parsing more robust

When validating file format, validateFileFormat (in write-file) analyzes the file signature (i.e., 'magic numbers') encoded as base64 rather than checking the 'data label.' This means the user no longer needs to include a data label when sending base64. Also, this means we no longer need a parser to strip the data label from base64 strings, so I removed the base64 parse file and related tests.

andrewvora reviewed (commented):
Nice change! Seems reliable. I would just add some basic tests that makes sure the validation still works

radiantdinosaurs committed:
Did some minor refactoring and added some tests and logging.

Added in more tests for the 'write-file' file. Did some minor refactoring here and there, nothing major. Added in winston to handle logging.

radiantdinosaurs committed:
Minor variable fix in write-file's unit tests.

andrewvora commented on app/utils/write_file/write-file.spec.js:
```diff
@@ -5,11 +5,36 @@ const proxyquire = require('proxyquire')
+const base64 = {
+    jpeg: '/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAMCAgICAgMCAgIDAwMDBAYEBAQEBAgGBgUGCQgKCgkICQkKDA8MCgsOCwkJDRENDg8QEBEQCgwSExIQEw8QEBD/2wBDAQMDAwQDBAgEBAgQCwkLEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBAQEBD/wgARCACAAIADAREAAhEBAxEB/8QAHAAAAgIDAQEAAAAAAAAAAAAABQYEBwIDCAAB/8QAGgEAAgMBAQAAAAAAAAAAAAAAAwQBAgUABv/aAAwDAQACEAMQAAAB6CaXhimucyMp6xtKVFQwAN/FsXYo3GDWabTxYTQ+v7uBBNynn6HaWllV ...
+    png: 'iVBORw0KGgoAAAANSUhEUgAAABgAAAAYCAYAAADgdz34AAAABGdBTUEAALGPC/xhBQAAAAFzUkdCAK7OHOkAAAAgY0hSTQAAeiYAAICEAAD6AAAAgOgAAHUwAADqYAAAOpgAABdwnLpRPAAAAAZiS0dEAP8A/wD/oL2nkwAABrdJREFUSMeFlktsXFcdxn/3nPuYO+PxjD0zHj/HtYuT1Gkbp2lITKUmbUUbCYRaiRVlVxawaRcsQEIsWFDBksAOsQQkJGhBCNKilkZp4jya1klwSpP4ESee8cSehz2emTtz7zmHhdNQ0Up8u7M455OO/t/3+1t8gb4yM0kYqZgj5VHL4iUprGO2FJNSWj0ASpmdSOklpc1pY3gjVOq8Y8vg3PzS5 ...
+}
```
I don't think you need the full PNG/JPEG. A little more than the magic number should be enough for the purposes of testing

andrewvora commented on app/utils/write_file/write-file.js:
```diff
@@ -55,19 +65,20 @@ async function writeJpegToDisk(base64, fileName) {
+                if (error) {
+                    logger.log('error', error)
+                    throw error.internalError()
+                }
```
`await` isn't actually doing anything here. You're treating the promise like a regular promise by using `then` and `catch`. The point of `await` is to synchronously run the promise object. You want to use an await try-catch approach here

```javascript
try { 
    const output = await pngToJpeg({quality: 90})(buffer)
    // the fs.writeFile piece must be promise-fied
    await writeFile(filename, output)
} catch (err) {
   if (err instanceof Error) {
   } else throw returnError.internalError()
}

return filename
```

```javascript
function writeFile(filename, output) {
   return new Promise((resolve, reject) => {
      fs.writeFile(filename, output, (error) => {
          if (error) {
                logger.log('error', error)
                reject(error.internalError())
          }
          else resolve()
      })
   })
}
```

This will roughly do what you're trying to accomplish. Syntax might need to be corrected for

andrewvora commented on app/utils/write_file/write-file.js:
```diff
@@ -35,13 +46,12 @@ const validateFileFormat = (base64) => {
+            if (error) {
+                logger.log('error', error)
+                throw returnError.internalError()
+            }
```
Same point here as in the `writePngToDisk` comment

andrewvora commented on app/utils/write_file/write-file.spec.js:
```diff
@@ -20,9 +45,52 @@ describe('write_file', () => {
+                })
+        })
+        it('checks that base64 is valid file format', (done) => {
+            mockFs.writeFile = () => Promise.resolve()
```
The places where you're mocking `writeFile` should be mocked in a way where you can "capture" and invoke the callback function. What you're doing works by coincidence, but it's not very obvious what behaviour you're mocking.

```javascript
mockFs.writeFile = (filename, content, callback) => {  callback(<error goes here>, <result goes here>) }
```

For your purposes, you just need to pass null into the callback, `callback(null)` to let it succeed. If you want it to fail, you call `callback(new Error())`

Do this everywhere you're mocking `fs.writeFile`

andrewvora reviewed (changes requested):
There's some more changes needed before everything is good to merge.

Before you start on any changes, I recommend doing the following:

1. Download SourceTree
2. Use it to rebase this branch on top of development, so you can get the latest changes.
3. Then make the changes

**Alternatively** if you're using WebForm, you should have a built in visual VCS manager. This might be easier. I encourage you to NOT do a merge

Since you'll be making multiple changes to the node app. It would be unadvised to keep doing that with the the all-in-one-docker-container approach

radiantdinosaurs committed:
Changed tests, added logging, and fixed some functions.

Changed the mock for fs and some variables here and there for write-file's test. Added error logging in tensorflow-client. Fixed await problems in write-file, as requested.

andrewvora reviewed (approved):
Alright, things look good to merge. Let's merge it and get started with building the frontend component. You'll get chances to revisit the backend as you extend functionality to serve the frontend
